File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    private int answer = 0;
    
    public int solution(int[] numbers, int target) {
        int[][] arr = new int[numbers.length][4];
        
        dfs(numbers[0] * 1, 1, target, numbers);
        dfs(numbers[0] * -1, 1, target, numbers);
        
        return answer;
    }
    
    public void dfs(int current, int idx, int target, int[] numbers) {
        if (idx == numbers.length && current == target) {
            answer++;
            return;
        }
        
        if (idx == numbers.length) return;
        
        dfs(current + numbers[idx], idx+1, target, numbers);
        dfs(current - numbers[idx], idx+1, target, numbers);
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.kt
class Solution {
    var answer = 0
    
    fun solution(numbers: IntArray, target: Int): Int {
        dfs(numbers, target, 0, 0, 0)
        
        return answer
    }
    
    fun dfs(
        numbers: IntArray,
        target: Int,
        count: Int,
        current: Int,
        idx: Int
    ) {
        if (count == numbers.size) {
            if (target == current) {
                answer++
            }
            
            return
        }
        
        val nextA = current + numbers[idx]
        val nextB = current - numbers[idx]
        
        dfs(numbers, target, count + 1, nextA, idx + 1)
        dfs(numbers, target, count + 1, nextB, idx + 1)
    }
}